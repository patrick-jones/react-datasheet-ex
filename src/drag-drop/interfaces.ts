import RDS from 'react-datasheet';
import {ConnectDragPreview, ConnectDragSource, ConnectDropTarget} from 'react-dnd';
import {ColumnHeader, HeaderRendererProps, RowRendererProps} from '../interfaces';

/**
 * Properties that must be provided to a drag and drop enabled row
 */
export interface DragDropRowProps<T extends RDS.Cell<T, V>, V = string> {
  /** The current row index */
  row: number;
  /** Callback for when a row is dropped on another */
  readonly onRowDropped: (source: number, target: number) => void;
}

/**
 * Properties that must be provided to a drag and drop enabled column header
 */
export interface DragDropHeaderProps {
  /** The current column index */
  col: number;
  /** Callback for one header is dropped on another */
  readonly onHeaderDropped: (source: number, target: number) => void;
}

/**
 * Drag source properties that are injected into a drag and drop provider
 */
export interface InjectedDragSourceProps {
  readonly connectDragSource: ConnectDragSource;
  readonly connectDragPreview: ConnectDragPreview;
}

/**
 * Drop target properties that are injected into a drag and drop provider
 */
export interface InjectedDropTargetProps {
  readonly connectDropTarget: ConnectDropTarget;
  readonly isDropOver: boolean;
}

/**
 * Combined drag source and drop target properties that are injected
 * into a drag and drop provider
 */
export interface InjectedDragDropProps extends InjectedDragSourceProps, InjectedDropTargetProps {}

/**
 * Extends `ReactDatasheet.RowRendererProps` to include
 * drag and drop properties
 */
export interface DragDropRowRendererProps<CellType, V = string>
  extends RowRendererProps<CellType, V>,
    DragDropRowProps<CellType, V> {}

/**
 * Extends `HeaderRendererProps` to include drag and drop properties.
 */
export interface DragDropHeaderRendererProps<T extends ColumnHeader = ColumnHeader>
  extends HeaderRendererProps<T>,
    DragDropHeaderProps {}
