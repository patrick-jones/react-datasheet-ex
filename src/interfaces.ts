import React, {CSSProperties} from 'react';
import RDS from 'react-datasheet';

export interface RowLocation {
  row: number;
}

export interface ColumnLocation {
  col: number;
}

export interface ColumnHeader {
  id: string;
  title: string;
  width?: number;
}

export interface HeaderRendererProps<T extends ColumnHeader = ColumnHeader> extends ColumnLocation {
  header: T;
}

export type HeaderRenderer<T extends ColumnHeader = ColumnHeader,
  R extends HeaderRendererProps<T> = HeaderRendererProps<T>> = React.ComponentType<R>;

export type HeaderContentRenderer<T extends ColumnHeader = ColumnHeader>
  = React.ComponentType<HeaderRendererProps<T>>;

export interface SheetRendererProps<
    T extends RDS.Cell<T, V>, V = string, H extends ColumnHeader = ColumnHeader
  >
  extends RDS.SheetRendererProps<T, V> {
  /** React style object that should be applied to the enclosing table */
  style?: CSSProperties | undefined;
  /**
   * An array of `ColumnHeader`, which are rendered into `th` elements
   * @default []
   */
  headers: H[];
  /**
   * Optional component or function to render the header cell contents.
   * The contents are rendered within a `th` element.
   * @default DefaultHeaderRenderer
   */
  headerRenderer: HeaderRenderer;
  /**
   * Optional component or function to render the header cell contents.
   * The contents are rendered within a `th` element.
   * @default DefaultHeaderContentRenderer
   */
  headerContentRenderer: HeaderContentRenderer;
}


export type ActionContentRenderer<T extends RowLocation = RowLocation>
  = React.ComponentType<T>;

export interface RowRendererProps<
    T extends RDS.Cell<T, V>, V = string, AP extends RowLocation = RowLocation
  > extends RDS.RowRendererProps<T, V> {

  actionContentRenderer?: ActionContentRenderer<AP>;
}
