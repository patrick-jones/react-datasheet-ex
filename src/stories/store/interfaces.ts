import RDS from 'react-datasheet';
import {ColumnHeader, Overflow} from '../../interfaces';

export type ExampleCellValue = string | number | null;
export interface ExampleCellType extends RDS.Cell<ExampleCellType> {
  value: ExampleCellValue;
}

export interface ExampleModel {
  id: string;
  headers: ExampleHeader[];
  rows: ExampleRow[];
  selected: RDS.Selection | null;
  overflow: Overflow;
}

export interface ExampleHeader extends ColumnHeader {
  hidden?: boolean;
}

export interface ExampleRow {
  id: string;
  selected?: boolean;
  data: ExampleCellType[];
}

