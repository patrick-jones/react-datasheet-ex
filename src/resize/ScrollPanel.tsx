import React, {PureComponent, ReactNode} from 'react';
import {SCROLL_PANEL} from '../ClassNames';


export interface ScrollPanelProps {
  width?: number;
  height?: number;
}


export default class ScrollPanel
  extends PureComponent<ScrollPanelProps> {

  render(): ReactNode {
    const {width, height} = this.props;
    const style = {
      width: width ? `${width}px` : undefined,
      height: height ? `${height}px` : undefined,
    };
    return (
      <span className={SCROLL_PANEL} style={style}>
        {this.props.children}
      </span>
    );
  }
}
