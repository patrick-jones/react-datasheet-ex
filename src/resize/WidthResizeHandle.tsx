import React, {PureComponent, ReactNode} from 'react';

import {RESIZING_WIDTH, WIDTH_RESIZE_HANDLE} from '../ClassNames';
import {ResizableHeaderProps} from './interfaces';


const initialState = {
  initialOffset: 0,
};


export default class WidthResizeHandle
  extends PureComponent<ResizableHeaderProps, typeof initialState> {

  static defaultProps = {
    cellTag: 'TH',
    minSize: 17,
  };

  state = initialState;

  elementRef = React.createRef<HTMLDivElement>();

  originalWidth = () => {
    const th = header(this.elementRef.current, this.props.cellTag);
    return th ? th.clientWidth : 0;
  };

  handleMouseDown = (e: React.MouseEvent<HTMLElement>) => {
    const {index, onResizeStart} = this.props;
    const originalSize = this.originalWidth();
    const initialOffset = originalSize - e.pageX;

    this.setState({initialOffset});
    window.addEventListener('mouseup', this.handleMouseUp);
    window.addEventListener('mousemove', this.handleMouseMoved);
    document.body.classList.add(RESIZING_WIDTH);
    onResizeStart && onResizeStart(index, originalSize);
  };

  handleMouseMoved = (e: MouseEvent) => {
    const {index, onResizing, minSize} = this.props;
    if (onResizing) {
      const {initialOffset} = this.state;
      const requestedSize = initialOffset + e.pageX;
      if (requestedSize >= minSize) {
        onResizing(index, requestedSize);
      }
    }
  };

  handleMouseUp = (e: MouseEvent) => {
    window.removeEventListener('mouseup', this.handleMouseUp);
    window.removeEventListener('mousemove', this.handleMouseMoved);
    document.body.classList.remove(RESIZING_WIDTH);
    const {index, onResizeEnd, minSize} = this.props;
    const {initialOffset} = this.state;
    this.setState(initialState);
    onResizeEnd(index, Math.max(initialOffset + e.pageX, minSize));
  };

  render(): ReactNode {
    const {className, style} = this.props;

    return (
      <span
        ref={this.elementRef}
        className={className ? `${WIDTH_RESIZE_HANDLE} ${className}` : WIDTH_RESIZE_HANDLE}
        style={style}
        onMouseDown={this.handleMouseDown}
      >
        {this.props.children}
      </span>
    );
  }
}


export function header(el: HTMLElement | null, selector: string): Element | null {
  if (!el) {
    return null;
  }

  const upper = selector.toUpperCase();
  if (el.closest) {
    return el.closest(upper);
  }


  let parent = el.parentElement;
  while (parent && parent.nodeType === Node.ELEMENT_NODE) {
    if (parent.tagName === upper) {
      return parent;
    }
    parent = parent.parentElement;
  }

  return null;
}
