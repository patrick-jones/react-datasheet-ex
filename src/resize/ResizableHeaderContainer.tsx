import React, {PureComponent, ReactNode} from 'react';
import {ResizableHeaderProps} from './interfaces';
import HeaderContainer from '../HeaderContainer';
import WidthResizeHandle from './WidthResizeHandle';


export default class ResizableHeaderContainer
  extends PureComponent<ResizableHeaderProps> {

  static defaultProps = {
    cellTag: 'TH',
    minSize: 17,
  };

  render(): ReactNode {
    const {children, ...rest} = this.props;
    return (
      <HeaderContainer>
        {children}
        <WidthResizeHandle {...rest} />
      </HeaderContainer>
    );
  }
}
