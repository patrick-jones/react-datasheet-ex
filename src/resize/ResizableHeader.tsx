import React, {PureComponent, ReactNode} from 'react';
import {ResizableHeaderProps} from './interfaces';
import {ColumnHeader, HeaderRendererProps} from '../interfaces';
import HeaderTitle from '../HeaderTitle';
import ResizableHeaderContainer from './ResizableHeaderContainer';


export interface ResizableHeaderRendererProps<T extends ColumnHeader = ColumnHeader>
  extends ResizableHeaderProps, HeaderRendererProps<T> {
}

export default class ResizableHeader
  extends PureComponent<ResizableHeaderRendererProps> {

  static defaultProps = {
    cellTag: 'TH',
    minSize: 17,
  };

  render(): ReactNode {
    const {index, onResizeStart, onResizing, onResizeEnd, cellTag, minSize, ...rest} = this.props;
    return (
      <ResizableHeaderContainer
        index={index}
        onResizeStart={onResizeStart}
        onResizing={onResizing}
        onResizeEnd={onResizeEnd}
        cellTag={cellTag}
        minSize={minSize}
      >
        <HeaderTitle {...rest} />
      </ResizableHeaderContainer>
    );
  }
}
