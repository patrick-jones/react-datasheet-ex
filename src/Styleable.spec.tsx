import React from 'react';
import 'jest';
import {shallow} from 'enzyme';
import toJson from 'enzyme-to-json';
import {HeaderContainer, HeaderTitle} from './index';
import {
  HEADER_CELL_CONTAINER,
  HEADER_CELL_TITLE,
  WIDTH_RESIZE_HANDLE,
} from './ClassNames';
import {Styleable} from './interfaces';
import {WidthResizeHandle} from './resize';


function suite(
  name: string, Component: React.ComponentClass<Styleable>, CLASS_NAME: string, extraProps: any) {

  it('Should render defaults properly', () => {
    const wrapper = shallow(<Component {...extraProps}>content</Component>);
    expect(wrapper.prop('className')).toEqual(CLASS_NAME);
    expect(wrapper.prop('style')).toBeUndefined();
  });

  it('Should render options properly', () => {
    const className = 'foo-bar';
    const style = {background: 'blue'};
    const wrapper = shallow(
      <Component className={className} style={style} {...extraProps}>content</Component>
    );

    expect(wrapper.prop('className')).toEqual(`${CLASS_NAME} ${className}`);
    expect(wrapper.prop('style')).toEqual(style);
  });

  it('matches snapshot', () => {
    const className = 'foo-bar';
    const style = {background: 'blue'};
    const wrapper = shallow(
      <Component className={className} style={style} {...extraProps}>content</Component>
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });

}


describe('Styleable', () => {
  describe.each([
    ['HeaderContainer', HeaderContainer, HEADER_CELL_CONTAINER, {}],
    [
      'HeaderTitle',
      HeaderTitle,
      `${HEADER_CELL_TITLE} ${HEADER_CELL_TITLE}--clip`,
      {col: 0, header: {id: 'abcd-puppy', title: 'ABCD'}},
    ],
    ['WidthResizeHandle', WidthResizeHandle, WIDTH_RESIZE_HANDLE, {}],
  ])('%s', suite);
});
