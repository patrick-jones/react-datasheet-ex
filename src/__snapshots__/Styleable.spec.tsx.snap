// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Styleable HeaderContainer matches snapshot 1`] = `
<span
  className="rdx-header-cell-container foo-bar"
  style={
    Object {
      "background": "blue",
    }
  }
>
  content
</span>
`;

exports[`Styleable HeaderTitle matches snapshot 1`] = `
<span
  className="rdx-header-cell-title rdx-header-cell-title--clip foo-bar"
  style={
    Object {
      "background": "blue",
    }
  }
>
  content
</span>
`;

exports[`Styleable WidthResizeHandle matches snapshot 1`] = `
<span
  className="rdx-width-resize-handle foo-bar"
  onMouseDown={[Function]}
  style={
    Object {
      "background": "blue",
    }
  }
>
  content
</span>
`;
